Beeper/BarcelonaMautrixIPC/Experimental/StartNewChat.swift

```diff
@@ -5,56 +5,74 @@
 //  Created by Eric Rabil on 4/4/22.
 //
 
-import Foundation
 import Barcelona
-import Swog
-import Sentry
+import Foundation
 import IMCore
+import Sentry
+import Swog
 
 public struct ResolveIdentifierCommand: Codable {
     public var identifier: String
 }
 
 public struct GUIDResponse: Codable {
     public var guid: String
-    
+
     public init(_ guid: String) {
         self.guid = guid
     }
 }
 
+enum ResolveIdentifierCommandError: Error {
+    case timeout
+}
+
 extension ResolveIdentifierCommand: Runnable {
     public func run(payload: IPCPayload, ipcChannel: MautrixIPCChannel) {
-        let semaphore = DispatchSemaphore(value: 0)
-        var retrievedGuid: GUIDResponse? = nil
+        Task {
+            var retrievedGuid: GUIDResponse? = nil
+
+            let result = try await withThrowingTaskGroup(of: ChatLocator.SenderGUIDResult.self) { group in
+                group.addTask {
+                    try await ChatLocator.senderGUID(for: identifier)
+                }
+                group.addTask {
+                    let timeout: UInt64 = 12
+                    try await Task.sleep(nanoseconds: timeout * 1_000_000_000)
+                    CLWarn(
+                        "ResolveIdentifier",
+                        "Resolving identifier for \(identifier) timed out in \(timeout) seconds"
+                    )
+                    throw ResolveIdentifierCommandError.timeout
+                }
+                let result = try await group.next()!
+                group.cancelAll()
+                return result
+            }
```

Comment:
This is a bit ugly right now, I will refactor it into a proper timeout method the next time I need it.